File: apps/syn-cli/src/syn_cli/commands/metrics.py
"""Metrics commands — aggregated workflow and session metrics."""

from __future__ import annotations

import typer
from rich.panel import Panel
from rich.table import Table

from syn_cli._output import console, format_cost, format_tokens
from syn_cli.commands._api_helpers import api_get, build_params

app = typer.Typer(
    name="metrics",
    help="View aggregated workflow and session metrics",
    no_args_is_help=True,
)


@app.command("show")
def show_metrics(
    workflow_id: str | None = typer.Option(None, "--workflow", "-w", help="Filter by workflow ID"),
) -> None:
    """Show aggregated metrics (optionally filtered by workflow)."""
    params = build_params(workflow_id=workflow_id)
    data = api_get("/metrics", params=params)

    panel_text = (
        f"[bold]Workflows:[/bold] {data.get('total_workflows', 0)} "
        f"(completed: {data.get('completed_workflows', 0)}, failed: {data.get('failed_workflows', 0)})\n"
        f"[bold]Sessions:[/bold] {data.get('total_sessions', 0)}\n"
        f"[bold]Tokens:[/bold] {format_tokens(data.get('total_tokens', 0))} "
        f"(in: {format_tokens(data.get('total_input_tokens', 0))}, "
        f"out: {format_tokens(data.get('total_output_tokens', 0))})\n"
        f"[bold]Cost:[/bold] {format_cost(data.get('total_cost_usd', '0'))}\n"
        f"[bold]Artifacts:[/bold] {data.get('total_artifacts', 0)}"
    )
    console.print(Panel(panel_text, title="[cyan]Metrics[/cyan]", border_style="cyan"))

    phases = data.get("phases", [])
    if phases:
        table = Table(title="Phase Metrics")
        table.add_column("Phase", style="cyan")
        table.add_column("Status")
        table.add_column("Tokens", justify="right")
        table.add_column("Cost", justify="right")
        table.add_column("Duration", justify="right")
        table.add_column("Artifacts", justify="right")

        for p in phases:
            table.add_row(
                p.get("phase_name", p.get("phase_id", "-")),
                p.get("status", "-"),
                format_tokens(p.get("total_tokens", 0)),
                format_cost(p.get("cost_usd", "0")),
                f"{p.get('duration_seconds', 0):.1f}s",
                str(p.get("artifact_count", 0)),
            )
        console.print(table)

File: apps/syn-cli-node/src/commands/metrics.ts
/**
 * Metrics commands — aggregated workflow and session metrics.
 * Port of apps/syn-cli/src/syn_cli/commands/metrics.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { api, unwrap } from "../client/typed.js";
import { print, printDim } from "../output/console.js";
import { style, BOLD, CYAN } from "../output/ansi.js";
import { formatCost, formatDuration, formatStatus, formatTokens } from "../output/format.js";
import { Table } from "../output/table.js";

const showCommand: CommandDef = {
  name: "show",
  description: "Show aggregated workflow and session metrics",
  options: {
    workflow: { type: "string", short: "w", description: "Filter by workflow ID" },
  },
  handler: async (parsed: ParsedArgs) => {
    const workflow = (parsed.values["workflow"] as string | undefined) ?? null;

    const d = unwrap(await api.GET("/metrics", {
      params: { query: { workflow_id: workflow } },
    }), "Fetch metrics");

    print(style("Aggregated Metrics", CYAN));
    print(`  ${style("Workflows:", BOLD)}     ${d.total_workflows}`);
    print(`  ${style("Sessions:", BOLD)}      ${d.total_sessions}`);
    print(`  ${style("Input tokens:", BOLD)}  ${formatTokens(d.total_input_tokens)}`);
    print(`  ${style("Output tokens:", BOLD)} ${formatTokens(d.total_output_tokens)}`);
    print(`  ${style("Total cost:", BOLD)}    ${formatCost(d.total_cost_usd)}`);
    print(`  ${style("Artifacts:", BOLD)}     ${d.total_artifacts}`);

    const phases = d.phases ?? [];
    if (phases.length > 0) {
      const table = new Table({ title: "Phase Metrics" });
      table.addColumn("Phase");
      table.addColumn("Status");
      table.addColumn("Tokens", { align: "right" });
      table.addColumn("Cost", { align: "right" });
      table.addColumn("Duration", { align: "right" });
      table.addColumn("Artifacts", { align: "right" });

      for (const ph of phases) {
        table.addRow(
          ph.phase_name,
          formatStatus(ph.status),
          formatTokens(ph.total_tokens),
          formatCost(ph.cost_usd),
          formatDuration(ph.duration_seconds * 1000),
          String(ph.artifact_count),
        );
      }
      table.print();
    }

    if (phases.length === 0 && !d.total_workflows) {
      printDim("No metrics data available.");
    }
  },
};

export const metricsGroup = new CommandGroup("metrics", "View aggregated workflow and session metrics");
metricsGroup.command(showCommand);
